### PR TITLE
.Net: BugFix Serializing and Deserializing ChatHistory with ToolCalling details Update OpenAI Connector to use FunctionToolCallsProperty and add ChatHistoryTests

### LINKED ISSUE #4336: .Net: ChatHistory is not serializable RC4
**Describe the bug**
When serializing and deserializing a ChatHistory object, 2 problems occur: tool_calls have a null content instead of an empty string, which gets rejected by the OpenAI API the next time it sees those messages, and something within the internal structure of the ChatHistory changes, prompting the error "Invalid parameter: messages with role 'tool' must be a response to a preceeding message with 'tool_calls'" even though the objects look the same before and after serialization

**To Reproduce**
Here is my demo:
```
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.Connectors.OpenAI;
using System.ComponentModel;
using System.Text.Json;

namespace SKTest
{
    public class DemoPlugin
    {
        [KernelFunction, Description("creates a special poem")]
        public string CreateSpecialPoem()
        {          
            return "ABCDE";
        }
    }
    internal class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello, World!");
            var builder = Kernel.CreateBuilder();

            builder.AddAzureOpenAIChatCompletion(
                     "gpt-4-turbo",                      // Azure OpenAI Deployment Name
                     "endpoint", // Azure OpenAI Endpoint
                     "key");      // Azure OpenAI Key
            
            builder.Plugins.AddFromType<DemoPlugin>();
           var kernel = builder.Build();
            AzureOpenAIChatCompletionService service = new AzureOpenAIChatCompletionService("gpt-4-turbo",                      // Azure OpenAI Deployment Name
                     "endpoint", // Azure OpenAI Endpoint
                     "key");      // Azure OpenAI Key
      OpenAIPromptExecutionSettings settings = new() { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
      ChatHistory history = new ChatHistory();

      history.AddUserMessage("Make me a special poem");
var historyBeforeJson = JsonSerializer.Serialize(history.ToList(), new JsonSerializerOptions { WriteIndented = true});

      ChatMessageContent result = await service.GetChatMessageContentAsync(history, settings, kernel);          
      Console.WriteLine(result.ToString());
      history.AddUserMessage("ok thank you");

      ChatMessageContent resultOriginalWorking = await service.GetChatMessageContentAsync(history, settings, kernel);

      //problem 2: after deserialization, something within the message structure triggers the problem 
      //Invalid parameter: messages with role 'tool' must be a response to a preceeding message with 'tool_calls'
      //even though the JSONs are the same

      var historyJson = JsonSerializer.Serialize<ChatHistory>(history, new JsonSerializerOptions { WriteIndented = true });
      var historyAfterSerialization = JsonSerializer.Deserialize<ChatHistory>(historyJson);

      //problem 1: a call_tools message has a null content, gets serialized as null, and can not be used again without fixing that
      foreach (var damagedMessage in historyAfterSerialization.Where(msg => msg.Content == null))
      {
          damagedMessage.Content = string.Empty;
      }

      var historyAfterSerializationJson = JsonSerializer.Serialize(historyAfterSerialization, new JsonSerializerOptions { WriteIndented = true });


      ChatMessageContent resultError = await service.GetChatMessageContentAsync(historyAfterSerialization, settings, kernel);
      Console.WriteLine(resultError.ToString());
}
}
```

**Expected behavior**
ChatHistory to be serializable, an easier access to the underlying ChatMessages to recreate the History including toolcalls, or a hint on how the ChatHistory is supposed to be persisted instead

**Platform**
RC4, .net8

### PR DESCRIPTION
**Pull request by RogerBarreto**
Serializing the ToolCalls was losing the details about Name and Arguments, and serializing this lack of content back was rendering ToolCalls as an invalid history message and giving the Issue Error.

Resolves #4336 

Next PR, moving the Integration Tests as Unit Tests
## Summary
This pull request updates the OpenAI Connector to use the FunctionToolCallsProperty metadata key instead of the ToolCallsProperty metadata key. This change allows for the use of a list of ChatCompletionsFunctionToolCall objects instead of ChatCompletionsToolCall objects. Additionally, a new integration test, ChatHistoryTests.cs, is added to the IntegrationTests/Connectors/OpenAI directory. The test file contains tests for the ChatHistory feature of the OpenAI Connector, ensuring that the feature returns the expected results and handles errors correctly.

## Changes
- Update ClientCore.cs to use FunctionToolCallsProperty instead of ToolCallsProperty
- Update OpenAIChatMessageContent.cs to use FunctionToolCallsProperty instead of ToolCallsProperty
- Add ChatHistoryTests.cs to IntegrationTests/Connectors/OpenAI directory
- Implement tests for ChatHistory feature of OpenAI Connector
- Dispose of logger and test output helper in ChatHistoryTests.cs

---
*Powered by [Microsoft Semantic Kernel](https://github.com/microsoft/semantic-kernel)*

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 1]
```diff
@@ -55,7 +56,7 @@ internal OpenAIChatMessageContent(AuthorRole role, string? content, string model
     /// <summary>
     /// A list of the tools called by the model.
     /// </summary>
-    public IReadOnlyList<ChatCompletionsToolCall> ToolCalls { get; }
+    public IReadOnlyList<ChatCompletionsFunctionToolCall> ToolCalls { get; }
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:59]: Isn't this going to be a problem when the Azure SDK adds other kinds of tools?

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 2]
```diff
@@ -82,7 +80,7 @@ public IReadOnlyList<OpenAIFunctionToolCall> GetOpenAIFunctionToolCalls()
     }
 
     private static IReadOnlyDictionary<string, object?>? CreateMetadataDictionary(
-        IReadOnlyList<ChatCompletionsToolCall> toolCalls,
+        IReadOnlyList<ChatCompletionsFunctionToolCall> toolCalls,
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:83]: I would have thought just changing the type stored into the metadata dictionary word be sufficient. Is it not?

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        history.AddUserMessage("Make me a special poem");
+        var historyBeforeJson = JsonSerializer.Serialize(history.ToList(), s_jsonOptionsCache);
+        var service = kernel.GetRequiredService<IChatCompletionService>();
+        ChatMessageContent result = await service.GetChatMessageContentAsync(history, settings, kernel);
+        history.AddUserMessage("Ok thank you");
+
+        ChatMessageContent resultOriginalWorking = await service.GetChatMessageContentAsync(history, settings, kernel);
+        var historyJson = JsonSerializer.Serialize(history, s_jsonOptionsCache);
+        var historyAfterSerialization = JsonSerializer.Deserialize<ChatHistory>(historyJson);
+        ChatMessageContent shouldNotThrow = await service.GetChatMessageContentAsync(historyAfterSerialization!, settings, kernel);
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:68]: nit:
```suggestion
        var exception = await Record.ExceptionAsync(() => service.GetChatMessageContentAsync(historyAfterSerialization!, settings, kernel));
        
        Assert.Null(exception);
```

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    private void ConfigureAzureOpenAIChatAsText(IKernelBuilder kernelBuilder)
+    {
+        var azureOpenAIConfiguration = this._configuration.GetSection("Planners:AzureOpenAI").Get<AzureOpenAIConfiguration>();
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:73]: Is this correct configuration for `ChatHistory` tests?

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 2 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:83]: I'm trying only that and see if it suffices.

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:73]: Yeah, I need the planners configuration because it uses the gpt-4-preview to work. Older models will break.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 5]
```diff
@@ -110,6 +116,7 @@ public IReadOnlyList<OpenAIFunctionToolCall> GetOpenAIFunctionToolCalls()
             }
 
             // Add the additional entry.
+            newDictionary.Add(FunctionToolCallsProperty, toolCalls.OfType<ChatCompletionsFunctionToolCall>().ToList());
```
Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: Are both necessary?

Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: ONe don't serialize the function call specific details, the other does.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 2 above)
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs]: If it doesn't, we'll need to do something else... I don't think we can change the public surface area like this.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 5 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: ONe way to resolve this later is to actually transform the ChatCompletionsToolCall object in a JsonElement and add it to the dictionary, or better approach.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 1 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs]: Yes, moved back to use the abstraction, but the serialization is lost. I added a specific metadata property for it but I still don't think this is the best approach, probably should just add a JsonElement to the dictionary object.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 6]
```diff
@@ -18,9 +19,14 @@ public sealed class OpenAIChatMessageContent : ChatMessageContent
...
-    /// Gets the metadata key for the <see cref="ChatCompletionsToolCall.Id"/> name property.
+    /// Gets the metadata key for the list of <see cref="ChatCompletionsFunctionToolCall"/>.
+    /// </summary>
+    public static string FunctionToolCallsProperty => $"{nameof(ChatResponseMessage)}.FunctionToolCalls";
+
+    /// <summary>
+    /// Gets the metadata key for the list of <see cref="ChatCompletionsToolCall"/>.
     /// </summary>
-    public static string ToolCallsProperty => $"{nameof(ChatResponseMessage)}.{nameof(ChatResponseMessage.ToolCalls)}";
+    public static string ToolCallsProperty => $"{nameof(ChatResponseMessage)}.ToolCalls";
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:29]: Curious why you changed this?

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 5 above)
Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: Serializing both is going to duplicate information. If we have to serialize non-function tool calls, can we filter it down to just the ones that aren't functions? For now that should be empty, so most of the time we don't need to add anything.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: Or just do the jsonelement thing now, as you suggest

Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: Moving to internal

Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:119]: I moved the properties to internal and I will be changing this behavior internally.

Looking currently in a definitive solution, I tried the JsonElement approach but it didn't work initially, will look at this in more detail.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 6 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs]: Copy paste. Reverted.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 7]
```diff
@@ -110,6 +116,7 @@ public IReadOnlyList<OpenAIFunctionToolCall> GetOpenAIFunctionToolCalls()
             }
 
             // Add the additional entry.
+            newDictionary.Add(FunctionToolCallsProperty, toolCalls.OfType<ChatCompletionsFunctionToolCall>().ToList());
             newDictionary.Add(ToolCallsProperty, toolCalls);
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:120]: Should we remove this? 

This should only contain the same data as what is stored under `FunctionToolCallsProperty`. We could just serialise the one type of tools we know about and extend this later to serialise other tools.

Comment (stephentoub) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:120]: Yeah, I think the only way this goes wrong is if the sdk sends us another kind of tool, which doesn't exist yet. Let's do the simple thing now that doesn't negatively impact the serialized state, and then subsequently follow up with a change to use something that gives us more control over what's serialized and how.

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIFunctionToolCall.cs [hunk 8]
```diff
@@ -176,5 +176,5 @@ internal static ChatCompletionsFunctionToolCall CreateChatCompletionsFunctionToo
         s_type?.SetValue(c, "function");
         return c;
     }
-    private static readonly PropertyInfo? s_type = typeof(ChatCompletionsToolCall).GetProperty("Type", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
+    private static readonly PropertyInfo? s_type = typeof(ChatCompletionsFunctionToolCall).GetProperty("Type", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public);
```
Comment (stephentoub) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIFunctionToolCall.cs:179]: Was this change necessary? I wouldn't expect it to be.

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    public class MyHttpMessageHandler : DelegatingHandler
+    {
+        public MyHttpMessageHandler()
+        {
+            this.InnerHandler = new HttpClientHandler();
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            string? requestContent = await request.Content!.ReadAsStringAsync(cancellationToken);
```
Comment (stephentoub) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:112]: What is the purpose of this?

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+
+    public void Dispose()
+    {
+        this.Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    ~ChatHistoryTests()
```
Comment (stephentoub) [inline dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:124]: This doesn't need to be finalizable

File: dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs (hunk 9 above)
Comment (RogerBarreto) [reply dotnet/src/IntegrationTests/Connectors/OpenAI/ChatHistoryTests.cs:115]: I used to understand how to request was being sent to the server, was important to spot the problem, I can remove it....

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 7 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs]: Removed the property

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIFunctionToolCall.cs (hunk 8 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIFunctionToolCall.cs]: Reverted

**Review by dmytrostruk — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]
